compilator/storage.go

```diff
@@ -0,0 +1,262 @@
+package compilator
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/SUSE/fissile/model"
+
+	"github.com/graymeta/stow"
+	"github.com/graymeta/stow/local"
+	"github.com/machinebox/progress"
+	"github.com/mholt/archiver"
+	"github.com/satori/go.uuid"
+	"gopkg.in/yaml.v2"
+
+	// support Azure storage
+	_ "github.com/graymeta/stow/azure"
+	// support Google storage
+	_ "github.com/graymeta/stow/google"
+	// support swift storage
+	_ "github.com/graymeta/stow/swift"
+	// support s3 storage
+	_ "github.com/graymeta/stow/s3"
+	// support oracle storage
+	_ "github.com/graymeta/stow/oracle"
+)
+
+// DownloadProgressEvent represents a delegate for updating progress when downloading
+type DownloadProgressEvent = func(progressPercentage float64)
+
+// PackageStorage represents a compiled BOSH package location
+type PackageStorage struct {
+	location           stow.Location
+	Kind               string
+	Config             stow.Config
+	CompilationWorkDir string
+	container          stow.Container
+	ImageName          string
+	ReadOnly           bool
+}
+
+type packageStorageConfig struct {
+	Kind          string `yaml:"boshPackageCacheKind"`
+	ReadOnly      bool   `yaml:"boshPackageCacheReadOnly"`
+	ContainerPath string `yaml:"boshPackageCacheLocation"`
+}
+
+// NewPackageStorageFromConfig creates a new PackageStorage based on a configuration file
+func NewPackageStorageFromConfig(configFilePath, compilationWorkDir, stemcellImageName string) (*PackageStorage, error) {
+	var packageCacheConfigReader []byte
+
+	// Read a yaml file that contains a stow configuration
+	if _, err := os.Stat(configFilePath); err != nil {
```

Comment:
This does not work because a literal config in JSON format may be longer than the maximum file name length:

```
stat {"boshPackageCacheKind":"s3","boshPackageCacheLocation":"package-cache","boshPackageCacheReadOnly":false,"access_key_id":"Q3AM3UQ867SPQQA43P2F","secret_key":"zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG","region":"us-east-1","endpoint":"minio.example.com:9000","disable_ssl":"true"}: file name too long
```

I would just get rid of the check (nobody sane would use a filename starting with a brace):

```
diff --git a/compilator/storage.go b/compilator/storage.go
index 52a6385..47a1519 100644
--- a/compilator/storage.go
+++ b/compilator/storage.go
@@ -55,18 +55,10 @@ type packageStorageConfig struct {
 // NewPackageStorageFromConfig creates a new PackageStorage based on a configuration file
 func NewPackageStorageFromConfig(configFilePath, compilationWorkDir, stemcellImageName string) (*PackageStorage, error) {
        var packageCacheConfigReader []byte
+       var err error

-       // Read a yaml file that contains a stow configuration
-       if _, err := os.Stat(configFilePath); err != nil {
-               if os.IsNotExist(err) {
-                       if strings.HasPrefix(configFilePath, "{") {
-                               packageCacheConfigReader = []byte(configFilePath)
-                       } else {
-                               return nil, nil
-                       }
-               } else {
-                       return nil, err
-               }
+       if strings.HasPrefix(configFilePath, "{") {
+               packageCacheConfigReader = []byte(configFilePath)
        } else {
                packageCacheConfigReader, err = ioutil.ReadFile(configFilePath)
                if err != nil {
@@ -77,10 +69,10 @@ func NewPackageStorageFromConfig(configFilePath, compilationWorkDir, stemcellIma
        var stowConfig map[string]interface{}
        var packageCacheConfig packageStorageConfig

-       if err := yaml.Unmarshal(packageCacheConfigReader, &stowConfig); err != nil {
+       if err = yaml.Unmarshal(packageCacheConfigReader, &stowConfig); err != nil {
                return nil, fmt.Errorf("Failed to unmarshal the package cache config file: %s", err.Error())
        }
-       if err := yaml.Unmarshal(packageCacheConfigReader, &packageCacheConfig); err != nil {
+       if err = yaml.Unmarshal(packageCacheConfigReader, &packageCacheConfig); err != nil {
                return nil, fmt.Errorf("Failed to unmarshal the package cache config file: %s", err.Error())
        }
```